### PR TITLE
DeclarationContext fixes

### PR DESCRIPTION
**Pull request by eregon**
* DefaultDefinee now captures the module to define methods on and no longer relies on LexicalScope and other approximations.
* Removes the `InternalMethod#capturedDefaultDefinee` hack.
* Fixes the definition of nested methods inside instance_eval/class_eval.

### CODE REVIEW

File: spec/ruby/core/module/define_method_spec.rb [hunk 1]
```diff
@@ -631,3 +631,28 @@ def m(a, b, *c)
...
+  describe "and a method is defined inside" do
+    it "defines the nested method in the default definee where the Proc was created" do
+      prc = nil
+      t = Class.new do
+        prc = -> {
+          def nested_method_in_proc_for_define_method
+            42
+          end
+        }
+      end
```
Comment (chrisseaton) [inline spec/ruby/core/module/define_method_spec.rb]: Some of these edge cases are insane.

Comment (eregon) [reply spec/ruby/core/module/define_method_spec.rb]: Indeed. And the default definee in Ruby can only be observed through where the method gets defined, which doesn't help in making tests/specs clear.

**Comment by eregon** [comment]
@bjfish Now I get the same failures as in the refinements PR, got to investigate them!

**Comment by eregon** [comment]
The fix was much simpler than expected: b8f45e8

**Pull request merged**